### PR TITLE
Add the ability to set a filter

### LINKED ISSUE #13: Is there a way to filter logs origin?
My project use multiple crates. I only want the logs from mine, not all the others.

### PR DESCRIPTION
**Pull request by paulrouget**
Fix #13.

Is using a lazy_static the best option here?

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -257,3 +271,9 @@ pub fn init_once(log_level: Level) {
         debug!("android_logger: log::set_logger failed: {}", err);
     }
 }
+
+/// Filter which modules print logs. Separated by comma.
+/// `None` will print all the logs.
+pub fn set_filter(filter: Option<String>) {
```
Comment (Nercury) [inline src/lib.rs]: Maybe use more specific name here, like 'set_module_path_filter'?

Comment (Nercury) [reply src/lib.rs]: Scratch that, I have better idea posted bellow.

**Comment by Nercury** [comment]
I was thinking about future extensibility a bit, and I think a better idea would be to move to initialization similar to that in env_logger. Of course, we don't need anything as complicated.

I think of changing the API to this:

```rust
    android_logger::init_once(
        Filter::default()
            .with_level(Level::Error)
            .with_module_name_filter("crate::name")
    )
```

Instead of using separate `lazy_static` flag, you can initialize whole AndroidLogger as lazy_static:

```rust
/// Underlying android logger, for cases where `init_once` abstraction is not enough.
pub struct AndroidLogger {
    filter: sync::RwLock<Filter>,
}

lazy_static! {
   static ref ANDROID_LOGGER: AndroidLogger = AndroidLogger::default();
}
```

Now this filter will be accessible inside the log function. The effect is the same, but it will be more flexible in the future. This if course breaks the API and I will need to update the docs. Not a big deal.

I played around a bit to check if this will work, and indeed it will. Example `init_once`:

```rust
pub fn init_once(filter: Filter) {
    if let Err(err) = log::set_logger(&*ANDROID_LOGGER) {
        debug!("android_logger: log::set_logger failed: {}", err);
    } else {
        if let Some(level) = filter.log_level {
            log::set_max_level(level.to_level_filter());
        }
        *ANDROID_LOGGER.filter.write().expect("failed to set log filter") = filter;
    }
}
```

The error is usually received when init_once is called the second time, so I think it is logical to move `log::set_max_level` to be called only on success, as well as swapping the filter struct.

And the filter would be like this:

```rust
pub struct Filter {
    log_level: Option<Level>,
    module_name_filter: Option<String>,
}

impl Default for Filter {
    fn default() -> Self {
        Filter {
            log_level: None,
            module_name_filter: None,
        }
    }
}

impl Filter {
    pub fn with_level(mut self, level: Level) -> Self {
        self.log_level = Some(level);
        self
    }

    pub fn with_module_name_filter(mut self, filter: &str) -> Self {
        self.module_name_filter = Some(filter.into());
        self
    }
}
```

Yeah, I almost did everything myself, but this is not tested, because I don't have an android project to try this out. But it builds ;)

**Comment by paulrouget** [comment]
Looks about right. I let you do a PR. I can give a try.

**Comment by paulrouget** [comment]
And thank you :)

**Comment by Nercury** [comment]
Continued in: https://github.com/Nercury/android_logger-rs/pull/15

**Pull request closed without merging**